### ISSUE TITLE
bug: readContract fails with data?.slice is not a function

### ISSUE THREAD

**Issue by denyshorman** [wagmi-dev/viem#484]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package Version

0.3.22

### Current Behavior

1. Call the contract `ownerOf` function using `publicClient.readContract` method
2. Expecting to catch `TokenDoesNotExist` custom contract error but receiving `ContractFunctionExecutionError`

```
ERROR ContractFunctionExecutionError: An unknown error occurred while executing the contract function "ownerOf".

Contract Call:
  address:   0x5FbDB2315678afecb367f032d93F642f64180aa3
  function:  ownerOf(uint256 token)
  args:             (67496899914374766384450646758818042943588521285977769413232173570159636609439)

Docs: https://viem.sh/docs/contract/readContract.html
Details: data?.slice is not a function
Version: viem@0.3.22
    at getContractError (getContractError.js:24:12)
    at readContract.js:25:31
    at Generator.throw (<anonymous>)
    at asyncGeneratorStep (asyncToGenerator.js:3:1)
    at _throw (asyncToGenerator.js:25:1)
    at _ZoneDelegate.invoke (zone.js:368:26)
    at Object.onInvoke (core.mjs:26074:33)
    at _ZoneDelegate.invoke (zone.js:367:52)
    at Zone.run (zone.js:127:43)
    at zone.js:1257:36
```

### Expected Behavior

It is expected to catch a custom error that is thrown by a smart contract.

### Steps To Reproduce

1. Create a smart contract 

```solidity
pragma solidity ^0.8.0;

contract TestContract {
    function ownerOf(uint256 token) public view returns (address) {
        revert TokenDoesNotExist(token);
    }
    
    error TokenDoesNotExist(uint256 token);
}
```

2. Deploy the smart contract to the local network (using Hardhat or other frameworks)

3. Call the function using viem library

```typescript
import { createPublicClient, http, parseAbi } from 'viem';

const client = createPublicClient({
    transport: http('http://localhost:8545'),
});

const abiString = [
    "function ownerOf(uint256 token) public view returns (address)",
    "error TokenDoesNotExist(uint256 token)",
];

const abi = parseAbi(abiString);

try {
    const owner = await client.readContract({
        address: '0x5FbDB2315678afecb367f032d93F642f64180aa3',
        abi,
        functionName: 'ownerOf',
        args: [67496899914374766384450646758818042943588521285977769413232173570159636609439n],
    });
} catch(e: any) {
    // should be TokenDoesNotExist
    // but receiving ContractFunctionExecutionError: An unknown error occurred while executing the contract function "ownerOf"
}
```

### Anything else?

HTTP Request Payload

```json
{"jsonrpc":"2.0","id":100,"method":"eth_call","params":[{"data":"0x6352211e9539e43b25b6dfdf9030a386321445083184e4d02807153275cd55d1207f799f","to":"0x5FbDB2315678afecb367f032d93F642f64180aa3"},"latest"]}
```

HTTP Response Payload
```json
{"jsonrpc":"2.0","id":100,"error":{"code":-32603,"message":"Error: VM Exception while processing transaction: reverted with custom error 'TokenDoesNotExist(67496899914374766384450646758818042943588521285977769413232173570159636609439)'","data":{"message":"Error: VM Exception while processing transaction: reverted with custom error 'TokenDoesNotExist(67496899914374766384450646758818042943588521285977769413232173570159636609439)'","data":"0xc927e5bf9539e43b25b6dfdf9030a386321445083184e4d02807153275cd55d1207f799f"}}}
```

![image](https://github.com/wagmi-dev/viem/assets/18450062/43f5ce1e-064f-4110-99c1-ab3206672816)

**Comment by iuriiiurevich** [comment]
This issue is related to the Hardhat's response format: https://github.com/NomicFoundation/hardhat/issues/3157
So it is `error: {message: "...", data: {data: "0x..." , message: "..."}}` instead of Viem's expected `error: {message: "...", data: "0x..."}`.

**Comment by jxom** [comment]
Ah, good catch. Fixed on 3a8966c85b12fc831c1756039fd7ec0c83cc093d.

**Comment by denyshorman** [comment]
Hey @jxom

Would you be able to take a look at this issue again?

I am reproducing two different errors calling the same method in two separate chains.

**Hardhat**

![image](https://github.com/wagmi-dev/viem/assets/18450062/2f0c762e-4212-454f-95e0-a35848d5dc62)

**Polygon**

![image](https://github.com/wagmi-dev/viem/assets/18450062/88bad37a-d8dc-4e0f-9ec9-5ade4c74f177)

Hardhat response
```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "error": {
    "code": -32603,
    "message": "Error: VM Exception while processing transaction: reverted with custom error 'TokenDoesNotExist(90941931887063947828948491733632283283969577934772334021012210786827295413179)'",
    "data": {
      "message": "Error: VM Exception while processing transaction: reverted with custom error 'TokenDoesNotExist(90941931887063947828948491733632283283969577934772334021012210786827295413179)'",
      "data": "0xc927e5bfc90f4f331be393a62fe49aa66e215d75c5175396030f85a801afba1af61653bb"
    }
  }
}
```

Polygon response
```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "error": {
    "code": 3,
    "message": "execution reverted",
    "data": "0xc927e5bfc90f4f331be393a62fe49aa66e215d75c5175396030f85a801afba1af61653bb"
  }
}
```

Would it be possible to somehow adapt to Hardhat's error format to have `ContractFunctionRevertedError` in the client code?
